device/src/efuse.rs

```diff
@@ -0,0 +1,198 @@
+
+        let kp = key_purpose as u8;
+        match key_number {
+            0 => buff[11] = kp,
+            1 => buff[11] = kp << 4,
+            2 => buff[12] = kp,
+            3 => buff[12] = kp << 4,
+            4 => buff[13] = kp,
+            5 => buff[13] = kp << 4,
+            _ => return Err(EfuseError::EfuseError),
+        }
+
+        // We bundle every config in Block 0 to minimize write operations
+        // Todo: Write key purpose and rw flags for multiple keys
+        // Disable write to key block
+        let mut write_disable = 0x01_u32 << (23 + key_number);
+        // Disable write to key purpose
+        write_disable += 0x01_u32 << (8 + key_number);
+        buff[0..4].copy_from_slice(&write_disable.to_le_bytes());
+
+        #[cfg(feature = "efuse_protect")]
+        self.set_read_protect(key_number, &mut buff);
+
+        self.write_block(&buff, 0)
+    }
+
+    /// # Safety
+    /// For disabling reading of efuse keys after efuse_debug
+    pub unsafe fn set_read_protect(&self, key_number: u8, buff: &mut [u8; 32]) {
+        // Disable read key
+        let read_disable = 0x01_u8 << key_number;
+        buff[4] = read_disable;
+    }
+
+    /// # Safety
+    unsafe fn write_block(&self, data: &[u8; 32], block_number: u8) -> Result<(), EfuseError> {
+        let efuse = &self.efuse;
+
+        let mut to_burn: [u32; 11] = [0; 11];
+
+        // Generate and write Reed-Solomon ECC
+        // Efuse controller ignores RS code for blocks 0 and 1
+        let rs_enc = reed_solomon::Encoder::new(12);
+        let ecc = rs_enc.encode(data);
+
+        // Flip efuse words to little endian
+        for (i, word) in ecc.chunks(4).enumerate() {
+            let n = u32::from_le_bytes(word.try_into().unwrap());
+            to_burn[i] = n;
+        }
+
+        // Write to efuse controller register
+        efuse.pgm_data0().write(|w| w.bits(to_burn[0]));
+        efuse.pgm_data1().write(|w| w.bits(to_burn[1]));
+        efuse.pgm_data2().write(|w| w.bits(to_burn[2]));
+        efuse.pgm_data3().write(|w| w.bits(to_burn[3]));
+        efuse.pgm_data4().write(|w| w.bits(to_burn[4]));
+        efuse.pgm_data5().write(|w| w.bits(to_burn[5]));
+        efuse.pgm_data6().write(|w| w.bits(to_burn[6]));
+        efuse.pgm_data7().write(|w| w.bits(to_burn[7]));
+
+        efuse.pgm_check_value0().write(|w| w.bits(to_burn[8]));
+        efuse.pgm_check_value1().write(|w| w.bits(to_burn[9]));
+        efuse.pgm_check_value2().write(|w| w.bits(to_burn[10]));
+
+        self.send_write_command(block_number);
+
+        self.update_read_registers();
+
+        if self.get_programming_error_record(block_number) {
+            Err(EfuseError::EfuseWriteError(block_number))
+        } else {
+            Ok(())
+        }
+    }
+
+    unsafe fn send_write_command(&self, block_number: u8) {
+        let efuse = &self.efuse;
+
+        // Send opcode, blknum and write command
+        efuse.conf().write(|w| w.op_code().bits(0x5A5A));
```

Comment:
Weird to me that the PAC doesn't put this into a constant somewhere. Can we put that this is in a read command constant.